File: 2015/R/d16.R
df <- data.frame(Sue = 0, children = NA, cats = NA, samoyeds = NA, pomeranians = NA,
                 akitas = NA, vizslas = NA, goldfish = NA, trees = NA,
                 cars = NA, perfumes = NA)

d <- data.table::rbindlist(lapply(readLines("../inputs/input_16.txt"), 
  function(x) {
    sx <- gsub(" ", "", strsplit(x, "[:,]")[[1]])
    sx[1] <- as.integer(gsub("Sue", "", sx[1]))
    row <- df[df$Sue == 0, ]
    row$Sue <- as.integer(sx[1])
    for (i in seq(2, length(sx), by = 2)) {
      row[[sx[i]]] <- as.integer(sx[i+1])
    }
    row
  }
))

ish <- function(x, y, func = `==`) {
  is.na(x) | func(x, y)
}

d <- d[ish(d$children, 3) & ish(d$samoyeds, 2) & ish(d$akitas, 0) & 
       ish(d$vizslas, 0) & ish(d$cars, 2) & ish(d$perfumes, 1), ]

c(d$Sue[ish(d$cats, 7) & ish(d$pomeranians, 3) & 
        ish(d$goldfish, 5) & ish(d$trees,3)],
  d$Sue[ish(d$cats, 7, `>`) & ish(d$pomeranians, 3, `<`) & 
        ish(d$goldfish, 5, `<`) & ish(d$trees,3, `>`)])

File: 2015/python/d16.py
criteria = {"children": 3, "cats": 7, "samoyeds": 2, "pomeranians": 3, \
            "akitas": 0, "vizslas": 0, "goldfish": 5, "trees": 3, \
            "cars": 2, "perfumes": 1}

def read(filename):
    sues = [None] * 501
    i = 0
    for line in open(filename, 'r'):
        str = line.replace('\n', '').replace('Sue ', '').replace(' ', '').\
                   replace(':', ',').split(',')
        sues[i] = str
        i += 1
    return sues

def part1(sues):
    for sue in sues:
       for thing in range(1, len(sue), 2):
           ok = True
           if not int(sue[thing + 1]) == criteria[sue[thing]]:
               ok = False
               break
       if ok:
           return sue[0]

def part2(sues):
    for sue in sues:
       for thing in range(1, len(sue), 2):
           ok = True
           if sue[thing] in {'cats', 'trees'}:
               if int(sue[thing + 1]) <= criteria[sue[thing]]:
                   ok = False
                   break
           elif sue[thing] in {'pomeranians', 'goldfish'}:
             if int(sue[thing + 1]) >= criteria[sue[thing]]:
                   ok = False
                   break
           elif not int(sue[thing + 1]) == criteria[sue[thing]]:
               ok = False
               break
       if ok:
           return sue[0]

sues = read("../inputs/input_16.txt")
print(part1(sues))
print(part2(sues))
